### PR TITLE
"RDS burst balances are depleted" adventure

### PR DESCRIPTION
**Pull request by tisnik**
# Description

Slides to be presented on demo.

## Type of change

- Documentation update

## Testing steps

N/A

## Checklist
* [ ] `make before_commit` passes
* [ ] updated documentation wherever necessary
* [ ] added or modified tests if necessary
* [ ] updated schemas and validators in [insights-data-schemas](https://github.com/RedHatInsights/insights-data-schemas) in case of input/output change

### CODE REVIEW

File: docs/presentations/DB-write-speed/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+## Step by step guide to view the slides
+
+This assumes that you have go compiler and git installed and on $PATH of your system.
+
+```
+go get golang.org/x/tools/cmd/present
```
Comment (dpensi) [inline docs/presentations/DB-write-speed/README.md:6]: https://go.dev/doc/go-get-install-deprecation uhmmm...

File: docs/presentations/DB-write-speed/db-write-speed.slide [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+.image write_speed_2.png
+
+
+
+## Possible improvements
+
+- vacuuming
+- cleanup old records
+- (minor) get rid of `.report` postfix in all tables
+
```
Comment (dpensi) [inline docs/presentations/DB-write-speed/db-write-speed.slide:269]: use EXPLAIN on the most frequent queries?

File: docs/presentations/DB-write-speed/benchmark.txt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+Input messages  Reports   Recommendation  Rule hit   Time to write  Write speed
+Kafka           DB        DB              DB
+
+ 10000          10000         70000         70000      4m 27s      37 wr/sec
```
Comment (dpensi) [inline docs/presentations/DB-write-speed/benchmark.txt:4]: what exactly is `wr/sec` ?

Comment (juandspy) [reply docs/presentations/DB-write-speed/benchmark.txt:4]: writes per second?

File: docs/presentations/DB-write-speed/README.md (hunk 1 above)
Comment (tisnik) [reply docs/presentations/DB-write-speed/README.md:6]: ah good!

File: docs/presentations/DB-write-speed/db-write-speed.slide (hunk 2 above)
Comment (tisnik) [reply docs/presentations/DB-write-speed/db-write-speed.slide:269]: yes I'm about it

File: docs/presentations/DB-write-speed/benchmark.txt (hunk 3 above)
Comment (tisnik) [reply docs/presentations/DB-write-speed/benchmark.txt:4]: yes writes per second

File: docs/presentations/DB-write-speed/db-write-speed.slide [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+- gp2 volumes
+- 5.4 million I/O credits at 3000 IOPS
+- 3000 IOPS for 30 minutes
+- once the burst is exhausted, it delivers only 100 IOPS
+- replenished at the rate 3 IOPS per GiB per second
+- during downtime, burst credit can be replenished
+- Burst/baseline performances
+    - burst performance is constant
+    - baseline performance increases linearly with volume size
+    - for volumes higher that 1TB, burst performance is irrelevant
```
Comment (juandspy) [inline docs/presentations/DB-write-speed/db-write-speed.slide:48]: "for volumes higher tha**n** 1TB, burst performance is irrelevant"

Comment (tisnik) [reply docs/presentations/DB-write-speed/db-write-speed.slide:48]: ah good catch!

File: docs/presentations/DB-write-speed/db-write-speed.slide [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+ Delete on recommendation  (cost=0.00..2794.57 rows=9 width=6) (actual time=9.539..9.542 rows=0 loops=1)
+   ->  Seq Scan on recommendation  (cost=0.00..2794.57 rows=9 width=6) (actual time=0.015..9.401 rows=5 loops=1)
+         Filter: ((cluster_id)::text = '94288c61-b7db-4d23-9f81-ee88d376970d'::text)
+         Rows Removed by Filter: 69995
+ Planning Time: 0.143 ms
+ Execution Time: 9.603 ms
+(6 rows)
+```
+
+- seq scan?!?!?!
```
Comment (juandspy) [inline docs/presentations/DB-write-speed/db-write-speed.slide:369]: does this mean that `cluster_id` is not an index in the `recommendation` table?

File: docs/presentations/DB-write-speed/db-write-speed.slide [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+```
+UPSERT INTO report table:         36 ms
+DELETE FROM recommendation table: 10 ms
+INSERT INTO recommendation table: 99 ms (10*9.983 ms)
+DELETE FROM rule_hit table:        0 ms
+INSERT INTO rule_hit table:       93 ms (10*9.363 ms)
+---------------------------------------
+                                 238 ms (4.2 wr/second)
+```
+
```
Comment (juandspy) [inline docs/presentations/DB-write-speed/db-write-speed.slide:474]: I suppose this scenarios also changes with the number of records in the DB?

File: docs/presentations/DB-write-speed/db-write-speed.slide [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+## Possible improvements
+
+- vacuuming
+- cleanup old records
+- proper index for `recommendation` table
+- (minor) get rid of `.report` postfix in all tables
+- make sure we use Amazon EBS SSD-based storage type
+- make sure that Notification DB does not affect Aggregator DB
+- `INSERT OR REPLACE INTO` (upsert) not needed, simple `INSERT` is enough
+- multiple INSERT statements into one statement (?)
```
Comment (juandspy) [inline docs/presentations/DB-write-speed/db-write-speed.slide:486]: Do you mean:
```
INSERT INTO table_name (column_list)
VALUES
    (value_list_1),
    (value_list_2),
    ...
    (value_list_n);
```
or using multiple `INSERT`?

**Review by juandspy — APPROVED** [review]

File: docs/presentations/DB-write-speed/db-write-speed.slide (hunk 5 above)
Comment (tisnik) [reply docs/presentations/DB-write-speed/db-write-speed.slide:369]: exactly :/

File: docs/presentations/DB-write-speed/db-write-speed.slide (hunk 6 above)
Comment (tisnik) [reply docs/presentations/DB-write-speed/db-write-speed.slide:474]: yes, it does. These numbers are for DB filled with ~ records we probably have in prod

File: docs/presentations/DB-write-speed/db-write-speed.slide (hunk 7 above)
Comment (tisnik) [reply docs/presentations/DB-write-speed/db-write-speed.slide:486]: the first one (but needs to be checked if help)

File: docs/presentations/DB-write-speed/db-write-speed.slide (hunk 5 above)
Comment (juandspy) [reply docs/presentations/DB-write-speed/db-write-speed.slide:369]: The index for that table is:
```
"recommendation_pk" PRIMARY KEY, btree (org_id, cluster_id, rule_fqdn, error_key)
```
This doesn't convert `cluster_id` into an index too?

File: docs/presentations/DB-write-speed/db-write-speed.slide (hunk 6 above)
Comment (juandspy) [reply docs/presentations/DB-write-speed/db-write-speed.slide:474]: nice. TY!

File: docs/presentations/DB-write-speed/db-write-speed.slide (hunk 7 above)
Comment (juandspy) [reply docs/presentations/DB-write-speed/db-write-speed.slide:486]: perfect

File: docs/presentations/DB-write-speed/db-write-speed.slide (hunk 5 above)
Comment (tisnik) [reply docs/presentations/DB-write-speed/db-write-speed.slide:369]: It creates just one "compound index" for the tuple `(org_id, cluster_id, rule_fqdn, error_key)`, but not just for `cluster_id` (at least AFAIK)

Comment (juandspy) [reply docs/presentations/DB-write-speed/db-write-speed.slide:369]: at least this is easy to test :)

**Pull request merged**